File: web/tests/supabase-api-headers.test.js
"use strict";

const assert = require("node:assert/strict");
const test = require("node:test");

const { supabaseApiHeaders } = require("../scripts/supabase-api-headers");

function legacyJwt(role) {
  const header = Buffer.from(JSON.stringify({ alg: "HS256", typ: "JWT" }))
    .toString("base64url");
  const payload = Buffer.from(JSON.stringify({ role }))
    .toString("base64url");
  return `${header}.${payload}.signature`;
}

test("legacy JWT API key keeps bearer compatibility", () => {
  const key = legacyJwt("anon");

  assert.deepEqual(supabaseApiHeaders(key), {
    apikey: key,
    Authorization: `Bearer ${key}`,
  });
});

test("publishable API key is not sent as a bearer token", () => {
  const key = "sb_publishable_public-test-key";

  assert.deepEqual(supabaseApiHeaders(key), {
    apikey: key,
  });
});

test("secret API key is rejected from public build clients", () => {
  assert.throws(
    () => supabaseApiHeaders("sb_secret_must-not-ship"),
    /secret API keys are not allowed/i,
  );
});

test("legacy service role JWT is rejected from public build clients", () => {
  assert.throws(
    () => supabaseApiHeaders(legacyJwt("service_role")),
    /service role API keys are not allowed/i,
  );
});

test("blank API key is rejected", () => {
  assert.throws(() => supabaseApiHeaders("   "), /API key is required/i);
});

File: shared/src/commonTest/kotlin/com/gallr/shared/data/network/SupabaseApiHeadersTest.kt
package com.gallr.shared.data.network

import io.ktor.http.HeadersBuilder
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNull

class SupabaseApiHeadersTest {
    @Test
    fun `legacy jwt api key is sent as apikey and bearer for compatibility`() {
        val apiKey = "eyJlegacy.header.signature"
        val headers = HeadersBuilder()

        headers.appendSupabaseApiKey(apiKey)

        assertEquals(apiKey, headers["apikey"])
        assertEquals("Bearer $apiKey", headers["Authorization"])
    }

    @Test
    fun `publishable api key is never sent as a bearer token`() {
        val apiKey = "sb_publishable_public-test-key"
        val headers = HeadersBuilder()

        headers.appendSupabaseApiKey(apiKey)

        assertEquals(apiKey, headers["apikey"])
        assertNull(headers["Authorization"])
    }

    @Test
    fun `secret api key is rejected from public clients`() {
        val headers = HeadersBuilder()

        assertFailsWith<IllegalArgumentException> {
            headers.appendSupabaseApiKey("sb_secret_must-not-ship")
        }
    }

    @Test
    fun `legacy service role jwt is rejected from public clients`() {
        val headers = HeadersBuilder()
        val serviceRoleJwt =
            "eyJhbGciOiJIUzI1NiJ9" +
                ".eyJyb2xlIjoic2VydmljZV9yb2xlIn0" +
                ".signature"

        assertFailsWith<IllegalArgumentException> {
            headers.appendSupabaseApiKey(serviceRoleJwt)
        }
    }

    @Test
    fun `blank api key is rejected before a request is built`() {
        val headers = HeadersBuilder()

        assertFailsWith<IllegalArgumentException> {
            headers.appendSupabaseApiKey("   ")
        }
    }
}
